malevich/interpreter/core.py

```diff
@@ -0,0 +1,150 @@
+import uuid
+from collections import defaultdict
+from typing import Any
+
+import jls_utils as jls
+
+from malevich._autoflow.tracer import tracer
+from malevich._utility.registry import Registry
+from malevich.constants import DEFAULT_CORE_HOST
+from malevich.interpreter.abstract import Interpreter
+from malevich.manifest import ManifestManager
+
+
+class CoreInterpreterState:
+    """State of the CoreInterpreter"""
+
+    def __init__(self) -> None:
+        # Involved operations
+        self.ops: dict[str, Any] = {}
+        # Dependencies (same keys as in self.ops)
+        self.depends = defaultdict(list[tuple[tracer, str]])
+        # Registry reference (just a shortcut)
+        self.reg = Registry()
+        # Manifest manager reference (just a shortcut)
+        self.manf = ManifestManager()
+        # Task configuration
+        self.cfg = jls.Cfg()
+        # Collections (key: operation_id, value: (`Collection` object, core_id,))
+        self.collections: dict[str, str] = {}
+        # Uploaded operations (key: operation_id, value: core_op_id)
+        self.core_ops: dict[str, str] = {}
+        # Interpreter parameters
+        self.params: dict[str, Any] = {}
+
+
+class CoreInterpreter(Interpreter[CoreInterpreterState, tuple[str, str]]):
+    """Inteprets the flow using Malevich Core API"""
+
+    def __init__(
+        self, core_host: str = DEFAULT_CORE_HOST, core_auth: tuple[str, str] = None
+    ) -> None:
+        super().__init__(CoreInterpreterState())
+        self.__core_host = core_host
+        self.__core_auth = core_auth
+        self._state.params["core_host"] = core_host
+        self._state.params["core_auth"] = core_auth
+
+        self.update_state()
+
+
+    def create_node(
+        self, state: CoreInterpreterState, tracer: tracer[tuple[str, dict]]
+    ) -> CoreInterpreterState:
+        state.ops[tracer.owner[0]] = {
+            **state.reg.get(tracer.owner[0], {}),
+            **tracer.owner[1],
+            "tracer": tracer,
+        }
+        return state
+
+    def create_dependency(
+        self,
+        state: CoreInterpreterState,
+        callee: tracer[tuple[str, dict]],
+        caller: tracer[tuple[str, dict]],
+        link: Any,  # noqa: ANN401
+    ) -> CoreInterpreterState:
+        state.depends[caller.owner[0]].append((callee.owner[0], link))
+        return state
+
+    def before_interpret(self, state: CoreInterpreterState) -> CoreInterpreterState:
+        jls.set_host_port(self.__core_host)
+        if self.__core_auth is None:
+            self.__core_auth = (uuid.uuid4().hex, uuid.uuid4().hex)
+            jls.create_user(self.__core_auth)
```

Comment:
It might be worth creating one fixed service account and tracking resources for it additionally locally (but this is an alternative, not urgent, and perhaps it needs to be at a lower level - we need to think about it)